yarn-project/simulator/src/avm/avm_machine_state.ts

```diff
@@ -52,6 +54,28 @@ export class AvmMachineState {
     return new AvmMachineState(state.l1GasLeft, state.l2GasLeft, state.daGasLeft);
   }
 
+  /**
+   * Consumes the given gas.
+   * Should any of the gas dimensions get depleted, it sets all gas left to zero and triggers
+   * an exceptional halt by throwing an OutOfGasError.
+   */
+  public consumeGas(gasCost: Partial<GasCost>) {
+    // Assert there is enough gas on every dimension.
+    const outOfGasDimensions = GasDimensions.filter(
+      dimension => this[`${dimension}Left`] - (gasCost[dimension] ?? 0) < 0,
+    );
+    // If not, trigger an exceptional halt.
+    // See https://yp-aztec.netlify.app/docs/public-vm/execution#gas-checks-and-tracking
+    if (outOfGasDimensions.length > 0) {
+      this.exceptionalHalt();
+      throw new OutOfGasError(outOfGasDimensions);
+    }
+    // Otherwise, charge the corresponding gas
+    for (const dimension of GasDimensions) {
+      this[`${dimension}Left`] -= gasCost[dimension] ?? 0;
```

Comment:
Are we losing any static type checking by using a string here or is TS smart enough?

Reply:
TS is smart enough! GasDimensions is defined `as const` so its type is not string but the actual values, and string literal types do the rest.